bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java

```diff
@@ -0,0 +1,298 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.novafinedust.internal.sds011protocol;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.TooManyListenersException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.util.HexUtils;
+import org.eclipse.smarthome.io.transport.serial.PortInUseException;
+import org.eclipse.smarthome.io.transport.serial.SerialPort;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEvent;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEventListener;
+import org.eclipse.smarthome.io.transport.serial.SerialPortIdentifier;
+import org.eclipse.smarthome.io.transport.serial.UnsupportedCommOperationException;
+import org.openhab.binding.novafinedust.internal.SDS011Handler;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.CommandMessage;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.Constants;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.ModeReply;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.SensorFirmwareReply;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.SensorMeasuredDataReply;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.SensorReply;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.SleepReply;
+import org.openhab.binding.novafinedust.internal.sds011protocol.messages.WorkingPeriodReply;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Central instance to communicate with the device, i.e. receive data from it and send commands to it
+ *
+ * @author Stefan Triller - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class SDS011Communicator implements SerialPortEventListener {
+
+    private final Logger logger = LoggerFactory.getLogger(SDS011Communicator.class);
+
+    private SerialPortIdentifier portId;
+    private SDS011Handler thingHandler;
+    private @NonNullByDefault({}) SerialPort serialPort;
+
+    private @Nullable OutputStream outputStream;
+    private @Nullable InputStream inputStream;
```

Comment:
If you want another way to get around the null checks. You could make these fields get populated from the constructor so you could make them non-null and final.
You would then turn your `initialize` method into a static factory method for `SDS011Communicator` instances.

Reply:
I'd rather structure my code in a way that makes it look clean, i.e. in this case having the serial port and the streams within one object, rather than splitting this throughout the code just to make an analysis tool happy.
I have made the null checks now...